### PR TITLE
BREAKING: Fix `InvalidOperationException` when using PropertiesDictionary in a multithreaded application and remove [Serializable] for it.

### PR DESCRIPTION
**Pull request by shaopeng-gh**
# Description:

Bug: `InvalidOperationException` when using PropertiesDictionary in a multithreaded application.
It is caused by use of Dictionary type instead of ConcurrentDictionary type.

# Fixes:

Change TypedPropertiesDictionary to use ConcurrentDictionary.
also found a existing issue with method SavePropertiesToXmlStream, fix by the way.

**BREAKING: removed [Serializable] for TypedPropertiesDictionary. 
[Serializable] is not for json and xml serialization, it is only for BinaryFormatter serialization. 
For SDK user that is currently using BinaryFormatter on TypedPropertiesDictionary object, 
will result in error: 
`SerializationException`: Type PropertiesDictionary is not marked as serializable**

# Error Details:

InvalidOperationException: Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.
at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
at Microsoft.CodeAnalysis.Sarif.PropertiesDictionary.TryGetProperty[T](String key, T& value)
at Microsoft.CodeAnalysis.Sarif.PropertiesDictionary.GetProperty[T](PerLanguageOption`1 setting, Boolean cacheDefault)
at Microsoft.CodeAnalysis.Sarif.PropertiesDictionary.GetProperty[T](PerLanguageOption`1 setting)
at Microsoft.CodeAnalysis.IL.Rules.EnableSpectreMitigations.AnalyzePortableExecutableAndPdb(BinaryAnalyzerContext context) in D:\a\r1\a\src\BinSkim.Rules\PERules\BA2024.EnableSpectreMitigations.cs:line 139
at Microsoft.CodeAnalysis.IL.Rules.WindowsBinaryAndPdbSkimmerBase.Analyze(BinaryAnalyzerContext context) in D:\a\r1\a\src\BinSkim.Rules\PERules\WindowsBinaryAndPdbSkimmerBase.cs:line 63
at Microsoft.CodeAnalysis.Sarif.Driver.MultithreadedAnalyzeCommandBase`2.AnalyzeTargetHelper(TContext context, IEnumerable`1 skimmers, ISet`1 disabledSkimmers)

### CODE REVIEW

File: src/Sarif/PropertiesDictionary.cs [hunk 1]
```diff
@@ -37,11 +36,6 @@ public PropertiesDictionary() : this(null)
         {
         }
 
-        protected PropertiesDictionary(SerializationInfo info, StreamingContext context)
-            : base(info, context)
-        {
-        }
```
Comment (eddynaka) [inline src/Sarif/PropertiesDictionary.cs:43]: check if the serialization will be the same comparing previous vs current

**Comment by eddynaka** [comment]
>                 SettingNameToDescriptionsMap = SettingNameToDescriptionsMap ?? new Dictionary<string, string>();

I think you might need to change this as well.
Be sure to test the cacheDescription part

---
Refers to: src/Sarif/TypedPropertiesDictionary.cs:72 in bef3fbe. [](commit_id = bef3fbeb4ad38813679891ee4e8c90dceec245c8, deletion_comment = False)

File: src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs [hunk 2]
```diff
@@ -124,6 +125,40 @@ public void PropertiesDictionary_RoundTripStringToVersionMap()
...
+                    taskList.Add(Task.Factory.StartNew(() =>
+                    properties.GetProperty(GetStringSetProperty(i))));
+                }
+
+                Task.WaitAll(taskList.ToArray());
+            });
+            Assert.Null(exception);
+        }
+
+
```
Comment (eddynaka) [inline src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs:161]: remove one line

File: src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs [hunk 3]
```diff
@@ -192,6 +227,10 @@ private PropertiesDictionary RoundTripThroughJson(PropertiesDictionary propertie
             new PerLanguageOption<StringSet>(
                 FEATURE, nameof(StringSetProperty), defaultValue: () => { return STRINGSET_DEFAULT; });
 
+        public static PerLanguageOption<StringSet> GetStringSetProperty(int i) =>
```
Comment (eddynaka) [inline src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs:230]: >GetStringSetProperty [](http://example.com/codeflow?start=51&length=20)

I would change to StringGetProperty

Comment (shaopeng-gh) [reply src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs:230]: StringSet is a actually class here, let me know 
how about change to "GenerateStringSetProperty"
the Get word may be confusing

**Comment by shaopeng-gh** [comment]
>                 SettingNameToDescriptionsMap = SettingNameToDescriptionsMap ?? new Dictionary<string, string>();

got it

---
In reply to: [989748904](https://github.com/microsoft/sarif-sdk/pull/2415#issuecomment-989748904) [](http://example.com/codeflow?ancestors=989748904,989748904)

---
Refers to: src/Sarif/TypedPropertiesDictionary.cs:72 in bef3fbe. [](commit_id = bef3fbeb4ad38813679891ee4e8c90dceec245c8, deletion_comment = False)

File: src/Sarif/PropertiesDictionary.cs (hunk 1 above)
Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionary.cs:43]: got it, thanks

**Comment by shaopeng-gh** [comment]
>                 SettingNameToDescriptionsMap = SettingNameToDescriptionsMap ?? new Dictionary<string, string>();

I think the cacheDescription part have problem before our changes. And we don't have a any test for that field.
I have made some changes, testing.

---
In reply to: [990377643](https://github.com/microsoft/sarif-sdk/pull/2415#issuecomment-990377643) [](http://example.com/codeflow?ancestors=990377643,989748904)

---
Refers to: src/Sarif/TypedPropertiesDictionary.cs:72 in bef3fbe. [](commit_id = bef3fbeb4ad38813679891ee4e8c90dceec245c8, deletion_comment = False)

File: src/Sarif/PropertiesDictionary.cs (hunk 1 above)
Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionary.cs:43]: the file written are the same

File: src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs (hunk 2 above)
Comment (shaopeng-gh) [reply src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs:194]: fixed

**Comment by shaopeng-gh** [comment]
>                 SettingNameToDescriptionsMap = SettingNameToDescriptionsMap ?? new Dictionary<string, string>();

added test for the cacheDescription save.

---
In reply to: [990734565](https://github.com/microsoft/sarif-sdk/pull/2415#issuecomment-990734565) [](http://example.com/codeflow?ancestors=990734565,989748904)

---
Refers to: src/Sarif/TypedPropertiesDictionary.cs:72 in bef3fbe. [](commit_id = bef3fbeb4ad38813679891ee4e8c90dceec245c8, deletion_comment = False)

File: src/Sarif/PropertiesDictionaryExtensionMethods.cs [hunk 4]
```diff
@@ -73,33 +74,28 @@ public static class PropertiesDictionaryExtensionMethods
                 StringSet stringSet = property as StringSet;
                 if (stringSet != null)
                 {
+                    SaveSettingNameToDescriptionMap(writer, key, settingNameToDescriptionMap);
```
Comment (eddynaka) [inline src/Sarif/PropertiesDictionaryExtensionMethods.cs:77]: >SaveSettingNameToDescriptionMap [](http://example.com/codeflow?start=20&length=31)

pls, add tests that cover this change.

File: src/Sarif/TypedPropertiesDictionary.cs [hunk 5]
```diff
@@ -13,7 +13,7 @@ public interface IMarker { }
 
     [Serializable]
     [JsonConverter(typeof(TypedPropertiesDictionaryConverter))]
-    public class TypedPropertiesDictionary<T> : Dictionary<string, T>, IMarker where T : new()
+    public class TypedPropertiesDictionary<T> : ConcurrentDictionary<string, T>, IMarker where T : new()
```
Comment (eddynaka) [inline src/Sarif/TypedPropertiesDictionary.cs:16]: >ConcurrentDictionary [](http://example.com/codeflow?start=48&length=20)

Update the title of the PR and add a new entry to the ReleaseHistory.

The title should be something like: "Fix InvalidOperationException when using PropertiesDictionary in a multithreaded application".

With that, if any customer checks the release history, the customer would know what motivated us to change this from Dictionary to ConcurrentDictionary.

This should also be applied to the PR description.

**Comment by michaelcfanning** [comment]
At a high-level, this PR isn't well explained. I will go annotate code changes that require more clarification.

File: src/Sarif/PropertiesDictionary.cs [hunk 6]
```diff
@@ -37,11 +37,6 @@ public PropertiesDictionary() : this(null)
         {
         }
 
-        protected PropertiesDictionary(SerializationInfo info, StreamingContext context)
```
Comment (michaelcfanning) [inline src/Sarif/PropertiesDictionary.cs:40]: >PropertiesDictionary [](http://example.com/codeflow?start=18&length=20)

Why did you remove this constructor? Preivously in .NET this ctor was required for all [Serializable] things, has that changed?

File: src/Sarif/ExtensionMethods.cs [hunk 7]
```diff
@@ -585,5 +586,15 @@ internal static void ConsumeElementOfDepth(this XmlReader xmlReader, int endElem
                 xmlReader.Read();
             }
         }
+
+        public static bool Add<T>(this ConcurrentDictionary<string, T> concurrentDictionary, string key, T value)
```
Comment (michaelcfanning) [inline src/Sarif/ExtensionMethods.cs:590]: >Add [](http://example.com/codeflow?start=27&length=3)

Why did you add this? What does the CurrentDictionary IDictionary.Add implementation call into? A non-thread-safe base method?

Comment (michaelcfanning) [inline src/Sarif/ExtensionMethods.cs:590]: >bool [](http://example.com/codeflow?start=22&length=4)

why does this extension method return a bool? IDictionary.Add returns void.

It's not a good idea to create a strange return value for 'Add', this will confuse experienced .NET programmers.

If you need to override 'Add', so that a concurrent dictionary behaves properly when it flows around to IDictionary-accepting things, this should return void. And in the case where Add would throw, you should consult the return value of TryAdd and throw on false.

File: src/Sarif/PropertiesDictionaryExtensionMethods.cs [hunk 8]
```diff
@@ -119,6 +115,19 @@ public static class PropertiesDictionaryExtensionMethods
             writer.WriteEndElement(); // Properties
         }
 
+        private static void SaveSettingNameToDescriptionMap(XmlWriter writer, string key, ConcurrentDictionary<string, string> settingNameToDescriptionMap)
```
Comment (michaelcfanning) [inline src/Sarif/PropertiesDictionaryExtensionMethods.cs:118]: >SaveSettingNameToDescriptionMap [](http://example.com/codeflow?start=28&length=31)

What are we doing here? Seems like you found a different issue and are oportunistically fixing it?

File: src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs [hunk 9]
```diff
@@ -124,6 +130,68 @@ public void PropertiesDictionary_RoundTripStringToVersionMap()
...
+                    properties.GetProperty(GenerateStringSetProperty(i))));
+                }
+
+                Task.WaitAll(taskList.ToArray());
+            });
+            Assert.Null(exception);
+        }
+
+        [Fact]
+        public void PropertiesDictionary_TestCacheDescription()
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs:167]: >PropertiesDictionary_TestCacheDescription [](http://example.com/codeflow?start=20&length=41)

Well done! Love seeing these tests relevant to your change, nice work.

File: src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs (hunk 3 above)
Comment (michaelcfanning) [reply src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs]: 'StringSet' is correct. :) I can completely understand why 'GenerateStringGetProperty' seems right, though! :)

**Review by michaelcfanning — CHANGES_REQUESTED** [review]
:clock1:

File: src/Sarif/ExtensionMethods.cs (hunk 7 above)
Comment (shaopeng-gh) [reply src/Sarif/ExtensionMethods.cs]: just checked the source code of CurrentDictionary,
the IDictionary<TKey, TValue>.Add implementation
did call the same method as the TryAdd
so we can use it thread safely.

Also agree that we should not make the Add has a return value people will get confused. 

updated the code.

Comment (shaopeng-gh) [reply src/Sarif/ExtensionMethods.cs]: agree and updated, see the other comment.

File: src/Sarif/PropertiesDictionaryExtensionMethods.cs (hunk 8 above)
Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionaryExtensionMethods.cs:118]: there were no tests for this and when I add tests this issue show up and blocking my test pass so I should fix it together.

avoid injecting four calls to your new method ---- agree, let me take a look

File: src/Sarif/PropertiesDictionaryExtensionMethods.cs (hunk 4 above)
Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionaryExtensionMethods.cs]: found this new method is not needed, removed.
the bug was comment not written to file,
the method I add was to test that. PropertiesDictionary_TestCacheDescription

File: src/Sarif/TypedPropertiesDictionary.cs (hunk 5 above)
Comment (shaopeng-gh) [reply src/Sarif/TypedPropertiesDictionary.cs:16]: done both, thanks

File: src/Sarif/PropertiesDictionaryExtensionMethods.cs (hunk 8 above)
Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionaryExtensionMethods.cs]: fixed to only one copy. the method is not needed.

File: src/Sarif/PropertiesDictionary.cs (hunk 6 above)
Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionary.cs:40]: the .net core concurrentdictionary is not binary serializable, so I removed it, 

question, do we even need it, it is not for json and xml serialization, it is only for BinaryFormatter serialization, in the solution I don't see code use BinaryFormatter

File: src/Sarif/TypedPropertiesDictionary.cs (hunk 5 above)
Comment (shaopeng-gh) [reply src/Sarif/TypedPropertiesDictionary.cs:16]: also added the ReleaseHistory for the other 2 PRs.

File: src/Sarif/PropertiesDictionary.cs (hunk 6 above)
Comment (michaelcfanning) [reply src/Sarif/PropertiesDictionary.cs:40]: Right, so if it's used anywhere it's in unit tests, I'd guess.

If you intend for this dictionary *not* to be serializable, you should remove the [Serializable] attribute. I think that's ok to do.

Comment (michaelcfanning) [reply src/Sarif/PropertiesDictionary.cs:40]: That's a good question and in future, definitely try to raise this before making this kind of change. Remember we are in an SDK, so it is possible that someone is using BinaryFormatter with this public type (even though we do not as you point out).

I think probably it is unlikely we're breaking someone out there. So this change is fine. *However* we need to very clearly mark this breaking change in behavior in the release notes.

Imagine someone using BInaryFormatter with this type, using an older SDK. Now they update to the new SDK and everything is broken. Ideally, your release notes would also mention the symptom these users would receive. Make sense? IOW, try to use BinaryFormatter again PropertiesDictionary (once you remove the [Serializable] attribute as well. See what exception is raised. Make sure the release notes reference this exception in case users experience it.

By doing this, we leave an easily followed trail of clues that tell developers what happened. That's the least we can do, given that we're introducing a breaking change. :)

**Comment by michaelcfanning** [comment]
>     [Serializable]

Let's remove this, if we're removing the serialization ctor.

---
Refers to: src/Sarif/PropertiesDictionary.cs:18 in 8eb304d. [](commit_id = 8eb304d1bee931243c956f8b83ee1264eb583948, deletion_comment = False)

**Comment by shaopeng-gh** [comment]
>     [Serializable]

fixed.

---
In reply to: [1004422690](https://github.com/microsoft/sarif-sdk/pull/2415#issuecomment-1004422690) [](http://example.com/codeflow?ancestors=1004422690,1004422690)

---
Refers to: src/Sarif/PropertiesDictionary.cs:18 in 8eb304d. [](commit_id = 8eb304d1bee931243c956f8b83ee1264eb583948, deletion_comment = False)

File: src/Sarif/PropertiesDictionary.cs (hunk 6 above)
Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionary.cs:40]: thanks for the great inputs, 
I have remove the [Serializable] attribute and ran it, the error is 
SerializationException: Type PropertiesDictionary is not marked as serializable
I have added it both in the PR desc and also the release history md file.

Comment (shaopeng-gh) [reply src/Sarif/PropertiesDictionary.cs:40]: @michaelcfanning 

additional question, do you want me to create a work item to check and remove all [Serializable] in the solution,
( I have not dig if any of them is used somewhere yet, just to bring up the discussion )

[https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide](https://docs.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide)

[https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2300](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2300)

File: src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs [hunk 10]
```diff
@@ -124,6 +130,69 @@ public void PropertiesDictionary_RoundTripStringToVersionMap()
             ((TypedPropertiesDictionary<Version>)properties[MapKey])[ValueKey].Should().Be(version);
         }
 
+        [Fact]
+        public void PropertiesDictionary_ConcurrentReadAndWrite()
+        {
+            var properties = new PropertiesDictionary();
+
+            Exception exception = Record.Exception(() =>
```
Comment (michaelcfanning) [inline src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs:138]: >Exception [](http://example.com/codeflow?start=41&length=9)

Did you verify that this test actually fails when the fix isn't applied?

Comment (shaopeng-gh) [reply src/Test.UnitTests.Sarif/PropertiesDictionaryTests.cs:138]: yes this test can repro the error before change the type.

File: src/Sarif/ExtensionMethods.cs [hunk 11]
```diff
@@ -585,5 +586,15 @@ internal static void ConsumeElementOfDepth(this XmlReader xmlReader, int endElem
                 xmlReader.Read();
             }
         }
+
+        public static void Add<T>(this ConcurrentDictionary<string, T> concurrentDictionary, string key, T value)
```
Comment (michaelcfanning) [inline src/Sarif/ExtensionMethods.cs:590]: >ConcurrentDictionary [](http://example.com/codeflow?start=39&length=20)

Why do we need these? Can't we just require our code to cast to IDictionary when appropriate?

Comment (michaelcfanning) [inline src/Sarif/ExtensionMethods.cs:590]: >Add [](http://example.com/codeflow?start=27&length=3)

Looking at this more closely, I think what we need to do is to reintroduce IDictionary<string, T> on the TypedPropertiesDictionary class, it is currently not available because it is implemented explicitly in ConcurrentDictionary.

**Review by michaelcfanning — APPROVED** [review]
:shipit:
